### PR TITLE
Filter events small plates

### PR DESCRIPTION
**Pull request by peterfpeterson**
This PR does a few small fixes to speed up event filtering. Using the script below, at a high level (on my laptop) `GenerateEventsFilter` goes from 14s to < 0.1s, and `FilterEvents` goes from 6s to 4s. This was done by using the [algorithm profiler](https://developer.mantidproject.org/AlgorithmProfiler.html) and linux's [perf tools](https://github.com/NAThompson/performance_tuning_tutorial).

**To test:**

Run the following before and after (part of a system test) and see that it is significantly faster
```python
from mantid.simpleapi import *
# add the system test data directory to the search path
# set the default facility to SNS
Load(Filename="HYS_13656-13658", OutputWorkspace="sum")
SetGoniometer(Workspace="sum", Axis0="s1,0,1,0,1")
GenerateEventsFilter(
    InputWorkspace="sum",
    OutputWorkspace="splboth",
    InformationWorkspace="info",
    UnitOfTime="Nanoseconds",
    LogName="s1",
    MaximumLogValue=90,
    LogValueInterval=1,
)
FilterEvents(
    InputWorkspace="sum",
    SplitterWorkspace="splboth",
    InformationWorkspace="info",
    FilterByPulseTime=True,
    GroupWorkspaces=True,
    OutputWorkspaceIndexedFrom1=True,
    OutputWorkspaceBaseName="split",
)
```

*There is no associated issue.*

*This does not require release notes* because is already covered by the event filtering release notes.

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 1]
```diff
@@ -1411,15 +1426,21 @@ void TimeSeriesProperty<TYPE>::create(const std::vector<Types::Core::DateAndTime
...
-  std::size_t num = new_values.size();
+  const std::size_t num = new_values.size();
+  m_values.reserve(num);
 
   m_propSortedFlag = TimeSeriesSortStatus::TSSORTED;
-  for (std::size_t i = 0; i < num; i++) {
-    TimeValueUnit<TYPE> newentry(new_times[i], new_values[i]);
-    m_values.emplace_back(newentry);
-    if (m_propSortedFlag == TimeSeriesSortStatus::TSSORTED && i > 0 && new_times[i - 1] > new_times[i]) {
+  // add the 0th value to simplify logic inside of the loop
```
Comment (jmborr) [inline Framework/Kernel/src/TimeSeriesProperty.cpp:1439]: I would split this into two:
```cpp
m_propSortedFlag = std::is_sorted(new_times) ? TimeSeriesSortStatus::TSSORTED : TimeSeriesSortStatus::TSUNSORTED;
for (std::size_t 0 = 1; i < num; i++)
    m_values.emplace_back(new_times[i], new_values[i]);
```
The reasons is that once you find the first violating `new_times[i - 1] > new_times[i]`, then there's no point in checking other `i-1, i` pairs.

File: Framework/Kernel/test/TimeSeriesPropertyTest.h [hunk 2]
```diff
@@ -1880,6 +1880,81 @@ class TimeSeriesPropertyTest : public CxxTest::TestSuite {
     return;
   }
 
+  void test_duplicateTimes() {
+    // emulate duplicates from the MDNorm test(s) with HYS_13656-13658
+    const DateAndTime time_first("2013-Jan-13 19:36:16.837000000");
+    const DateAndTime time_last("2013-Jan-13 19:36:21.900000202");
+    TimeSeriesProperty<double> prop("s1");
+    prop.addValue(time_first, -0.001722);
```
Comment (jmborr) [inline Framework/Kernel/test/TimeSeriesPropertyTest.h:1888]: the joys of `c++`

File: Framework/Kernel/src/TimeSeriesProperty.cpp (hunk 1 above)
Comment (peterfpeterson) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:1439]: I wasn't sure if `std:::sort` would be faster or if checking the order while looping through. The `if` on L1443 is checking that the flag is sorted first which should be cheap (enum comparison). I tried it out and we get another second out of the python script in the PR description.

**Review by jmborr — APPROVED** [review]
:ok_hand:

**Pull request merged**